src/librustc_resolve/lib.rs

```diff
@@ -3457,6 +3508,107 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         found_traits
     }
 
+    /// When name resolution fails, this method can be used to look up candidate
+    /// entities with the expected name. It allows filtering them using the
+    /// supplied predicate (which should be used to only accept the types of
+    /// definitions expected e.g. traits). The lookup spans across all crates.
+    ///
+    /// NOTE: The method does not look into imports, so it ignore all private-
+    /// made-public entities, such as the following code pattern:
+    /// ```
+    /// pub mod foo {
+    ///     mod bar { trait T {}; };
+    ///     // make private T accessible through an import:
+    ///     pub use bar::T as T;
+    /// }
```

Comment:
No problem, I probably should have mentioned this earlier :)
`pub use` statements actually can "reveal" private parts of a crate, for example
```rust
mod foo { pub trait T {} }
pub use foo::T; // We wouldn't be able to do this if `trait T` were not `pub`
```
Here, `T` would not be accessible to other traits if we didn't `pub use foo::T`.
What I mean by not distinguishing between public imports and public items is that when we load this external crate, both `binding`s for `T` will be `!binding.is_import()` -- that is, we "forget" if bindings came from items or from imports. In other words, we wouldn't be able to distinguish the above external crate from
```rust
mod foo { pub use T; }
pub trait T;
```